### PR TITLE
test: create a utility to test webhooks

### PR DESCRIPTION
**Pull request by woyuen**
This utility is exposed to plugins that can use it to perform webhook tests.

### CODE REVIEW

File: test/webhook-manager.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+const { getAccessTokenHeaders } = require('./auth')
+
+const noop = () => Promise.resolve()
+
+// Utility to test external service webhooks integration into Stelace
+// Two modes of functioning:
+// - test webhooks for real via Ngrok
+// - simulate webhooks by fetching and creating events
+//    (suitable if webhooks cannot be created via API and one cannot have access to Ngrok paid plan
+//     for reserved subdomains)
```
Comment (abruere) [inline test/webhook-manager.js]: I would be more precise and make the distinction between external events and Stelace events:

```js
// Two operating modes:
// - Test external webhook integration for real via Ngrok,
//     as long as the external service lets you create webhooks with some API,
//     or you have a paid ngrok plan and a stable URL to use for a longlived test webhook
// - Simulate external webhook by polling external service events
//    and propagate them to some Stelace endpoint, provided the external service
//    has some Event API to fetch latest external events.
```

File: test/webhook-manager.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+   *   poll events from external service and hit Stelace webhook endpoint
+   *   should return the last event timestamp for next fetch
+   *   this timestamp will be injected at the next call of the function `simulateWebhookSince`
+   * @example
+   *   const processedEventIds = {}
+   *
+   *   async simulateWebhookSince(lastFetchTimestamp) {
+   *     // we recommend to apply deduplication on events
+   *     // because events created at the same second can be not available at the same time
+   *     // via Events API
```
Comment (abruere) [inline test/webhook-manager.js]: ``` // via external Events API```

And this comment should go with a line having `processedEventIds` below.

File: test/webhook-manager.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+   *     const events = await exampleSdk.events.list({ gt: lastFetchTimestamp })
```
Comment (abruere) [inline test/webhook-manager.js]: ```externalServiceSdk.events…```

File: test/webhook-manager.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+   *         .post(webhookUrl)
+   *         .send(event)
+   *         .set({ authorization: `Basic ${basicAuth}` })
+   *         .expect(200)
+   *
+   *       processedEventIds[event.id] = true
+   *     }
+   *
+   *     const timestamps = events.map(e => e.created)
+   *     const maxTimestamp = _.max(timestamps) - 1 // to be sure to fetch processed events
```
Comment (abruere) [inline test/webhook-manager.js]: `any unprocessed events`?

And `1` is like a magic number here. Is it supposed to be milliseconds, seconds? Maybe you need some variable name for this safety margin.

I know it’s just an example but this is also a reference for implementation so this has to be easy to grasp.

File: test/webhook-manager.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+   *     return maxTimestamp ? maxTimestamp - 1 : null
+   *   }
+   *
+   * @param {Boolean}  [params.webhookFetchInterval = 2000] - specify the duration in milliseconds for events fetching
```
Comment (abruere) [inline test/webhook-manager.js]: `interval in milliseconds between calls to external Events API…`

File: test/webhook-manager.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    simulateWebhookSince,
+    webhookFetchInterval = 2000,
+    createWebhook,
+    removeWebhook,
+    tunnel
+  }) {
+    this.t = t
+
+    // minus one second to handle cases events are generated during the same second of webhook manager creation
+    this.lastEventTimestamp = getTimestamp() - 1
```
Comment (abruere) [inline test/webhook-manager.js]: Same as above: how can you tell it’s seconds and not milliseconds? It depends on the external service.

File: test/webhook-manager.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+      clearInterval(this.simulateWebhookInterval)
+    } else {
+      // https://github.com/bubenshchykov/ngrok#disconnect
+      await ngrok.disconnect(this.tunnelUrl)
+    }
+  }
+
+  // if isWebhookSimulated is false, really wait webhooks events
+  // otherwise, fetch events not retrieved from last time to simulate a real webhook running
+  async waitForEvents (waitDurationMs = 10000) {
```
Comment (abruere) [inline test/webhook-manager.js]: It seems it’s just some utility function exposed and comment above is more general advice about using `WehbookManager` than documentation about this function. In which case you should rewrite this comment and move it if needed.

File: test/webhook-manager.js [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+   *     }
+   *
+   *     const timestamps = events.map(e => e.created)
+   *     const maxTimestamp = _.max(timestamps) - 1 // to be sure to fetch processed events
+   *     return maxTimestamp ? maxTimestamp - 1 : null
+   *   }
+   *
+   * @param {Boolean}  [params.webhookFetchInterval = 2000] - specify the duration in milliseconds for events fetching
+   *   if isWebhookSimulated is true
+   * @param {Function} [params.createWebhook(tunnelUrl)] - can be provided if isWebhookSimulated is false
```
Comment (abruere) [inline test/webhook-manager.js]: You could add: `tunnelUrl being the ngrok endpoint to prepend to Stelace API endpoint path.`

I would also provide a full `@example` of `WebhookManager` use with an external webhook (non-simulated).

File: test/webhook-manager.js (hunk 6 above)
Comment (woyuen) [reply test/webhook-manager.js]: Indeed. At initialization, this value is set to `null` now.
The implementation of `simulateWebhookSince` should use the time unit that is easier to use in events list filter.

File: test/webhook-manager.js (hunk 7 above)
Comment (woyuen) [reply test/webhook-manager.js]: Good catch. This comment is obsolete and should be removed.

**Comment by woyuen** [comment]
In fact I realized the security margin was deduced twice (commit before using a variable to name the magic number), so I removed one of subtractions.
https://github.com/stelace/stelace/blob/eb06b4671efe0afac6a1be33df608ae7d6a5a4ff/test/webhook-manager.js#L46-L47

**Comment by abruere** [comment]
Yeah but no reason to remove them both.

LGTM.

**Pull request merged**